// @mysten/sui-grpc/dist/esm/proto/sui/rpc/v2beta2/argument.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
var Argument_ArgumentKind = /* @__PURE__ */ ((Argument_ArgumentKind2) => {
  Argument_ArgumentKind2[Argument_ArgumentKind2["ARGUMENT_KIND_UNKNOWN"] = 0] = "ARGUMENT_KIND_UNKNOWN";
  Argument_ArgumentKind2[Argument_ArgumentKind2["GAS"] = 1] = "GAS";
  Argument_ArgumentKind2[Argument_ArgumentKind2["INPUT"] = 2] = "INPUT";
  Argument_ArgumentKind2[Argument_ArgumentKind2["RESULT"] = 3] = "RESULT";
  return Argument_ArgumentKind2;
})(Argument_ArgumentKind || {});
class Argument$Type extends MessageType {
  constructor() {
    super("sui.rpc.v2beta2.Argument", [
      {
        no: 1,
        name: "kind",
        kind: "enum",
        opt: true,
        T: () => ["sui.rpc.v2beta2.Argument.ArgumentKind", Argument_ArgumentKind]
      },
      {
        no: 2,
        name: "input",
        kind: "scalar",
        opt: true,
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "result",
        kind: "scalar",
        opt: true,
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 4,
        name: "subresult",
        kind: "scalar",
        opt: true,
        T: 13
        /*ScalarType.UINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0) reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional sui.rpc.v2beta2.Argument.ArgumentKind kind */
        1:
          message.kind = reader.int32();
          break;
        case /* optional uint32 input */
        2:
          message.input = reader.uint32();
          break;
        case /* optional uint32 result */
        3:
          message.result = reader.uint32();
          break;
        case /* optional uint32 subresult */
        4:
          message.subresult = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(
              `Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`
            );
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(
              this.typeName,
              message,
              fieldNo,
              wireType,
              d
            );
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.kind !== void 0) writer.tag(1, WireType.Varint).int32(message.kind);
    if (message.input !== void 0) writer.tag(2, WireType.Varint).uint32(message.input);
    if (message.result !== void 0) writer.tag(3, WireType.Varint).uint32(message.result);
    if (message.subresult !== void 0) writer.tag(4, WireType.Varint).uint32(message.subresult);
    let u = options.writeUnknownFields;
    if (u !== false) (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Argument = new Argument$Type();
export {
  Argument,
  Argument_ArgumentKind
};
