// @mysten/sui-grpc/dist/esm/proto/sui/rpc/v2beta2/input.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Value } from "../../../google/protobuf/struct.js";
var Input_InputKind = /* @__PURE__ */ ((Input_InputKind2) => {
  Input_InputKind2[Input_InputKind2["INPUT_KIND_UNKNOWN"] = 0] = "INPUT_KIND_UNKNOWN";
  Input_InputKind2[Input_InputKind2["PURE"] = 1] = "PURE";
  Input_InputKind2[Input_InputKind2["IMMUTABLE_OR_OWNED"] = 2] = "IMMUTABLE_OR_OWNED";
  Input_InputKind2[Input_InputKind2["SHARED"] = 3] = "SHARED";
  Input_InputKind2[Input_InputKind2["RECEIVING"] = 4] = "RECEIVING";
  return Input_InputKind2;
})(Input_InputKind || {});
class Input$Type extends MessageType {
  constructor() {
    super("sui.rpc.v2beta2.Input", [
      {
        no: 1,
        name: "kind",
        kind: "enum",
        opt: true,
        T: () => ["sui.rpc.v2beta2.Input.InputKind", Input_InputKind]
      },
      {
        no: 2,
        name: "pure",
        kind: "scalar",
        opt: true,
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 3,
        name: "object_id",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "version",
        kind: "scalar",
        opt: true,
        T: 4,
        L: 0
      },
      {
        no: 5,
        name: "digest",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "mutable",
        kind: "scalar",
        opt: true,
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 1e3, name: "literal", kind: "message", T: () => Value }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0) reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional sui.rpc.v2beta2.Input.InputKind kind */
        1:
          message.kind = reader.int32();
          break;
        case /* optional bytes pure */
        2:
          message.pure = reader.bytes();
          break;
        case /* optional string object_id */
        3:
          message.objectId = reader.string();
          break;
        case /* optional uint64 version */
        4:
          message.version = reader.uint64().toBigInt();
          break;
        case /* optional string digest */
        5:
          message.digest = reader.string();
          break;
        case /* optional bool mutable */
        6:
          message.mutable = reader.bool();
          break;
        case /* optional google.protobuf.Value literal */
        1e3:
          message.literal = Value.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.literal
          );
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(
              `Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`
            );
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(
              this.typeName,
              message,
              fieldNo,
              wireType,
              d
            );
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.kind !== void 0) writer.tag(1, WireType.Varint).int32(message.kind);
    if (message.pure !== void 0) writer.tag(2, WireType.LengthDelimited).bytes(message.pure);
    if (message.objectId !== void 0)
      writer.tag(3, WireType.LengthDelimited).string(message.objectId);
    if (message.version !== void 0) writer.tag(4, WireType.Varint).uint64(message.version);
    if (message.digest !== void 0)
      writer.tag(5, WireType.LengthDelimited).string(message.digest);
    if (message.mutable !== void 0) writer.tag(6, WireType.Varint).bool(message.mutable);
    if (message.literal)
      Value.internalBinaryWrite(
        message.literal,
        writer.tag(1e3, WireType.LengthDelimited).fork(),
        options
      ).join();
    let u = options.writeUnknownFields;
    if (u !== false) (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Input = new Input$Type();
export {
  Input,
  Input_InputKind
};
