// @mysten/sui-grpc/dist/esm/proto/sui/rpc/v2beta2/owner.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
var Owner_OwnerKind = /* @__PURE__ */ ((Owner_OwnerKind2) => {
  Owner_OwnerKind2[Owner_OwnerKind2["OWNER_KIND_UNKNOWN"] = 0] = "OWNER_KIND_UNKNOWN";
  Owner_OwnerKind2[Owner_OwnerKind2["ADDRESS"] = 1] = "ADDRESS";
  Owner_OwnerKind2[Owner_OwnerKind2["OBJECT"] = 2] = "OBJECT";
  Owner_OwnerKind2[Owner_OwnerKind2["SHARED"] = 3] = "SHARED";
  Owner_OwnerKind2[Owner_OwnerKind2["IMMUTABLE"] = 4] = "IMMUTABLE";
  Owner_OwnerKind2[Owner_OwnerKind2["CONSENSUS_ADDRESS"] = 5] = "CONSENSUS_ADDRESS";
  return Owner_OwnerKind2;
})(Owner_OwnerKind || {});
class Owner$Type extends MessageType {
  constructor() {
    super("sui.rpc.v2beta2.Owner", [
      {
        no: 1,
        name: "kind",
        kind: "enum",
        opt: true,
        T: () => ["sui.rpc.v2beta2.Owner.OwnerKind", Owner_OwnerKind]
      },
      {
        no: 2,
        name: "address",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "version",
        kind: "scalar",
        opt: true,
        T: 4,
        L: 0
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0) reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional sui.rpc.v2beta2.Owner.OwnerKind kind */
        1:
          message.kind = reader.int32();
          break;
        case /* optional string address */
        2:
          message.address = reader.string();
          break;
        case /* optional uint64 version */
        3:
          message.version = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(
              `Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`
            );
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(
              this.typeName,
              message,
              fieldNo,
              wireType,
              d
            );
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.kind !== void 0) writer.tag(1, WireType.Varint).int32(message.kind);
    if (message.address !== void 0)
      writer.tag(2, WireType.LengthDelimited).string(message.address);
    if (message.version !== void 0) writer.tag(3, WireType.Varint).uint64(message.version);
    let u = options.writeUnknownFields;
    if (u !== false) (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Owner = new Owner$Type();
export {
  Owner,
  Owner_OwnerKind
};
