// @mysten/sui-grpc/dist/cjs/proto/sui/rpc/v2beta2/checkpoint.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var checkpoint_exports = {};
__export(checkpoint_exports, {
  Checkpoint: () => Checkpoint
});
module.exports = __toCommonJS(checkpoint_exports);
var import_runtime = require("@protobuf-ts/runtime");
var import_runtime2 = require("@protobuf-ts/runtime");
var import_runtime3 = require("@protobuf-ts/runtime");
var import_runtime4 = require("@protobuf-ts/runtime");
var import_executed_transaction = require("./executed_transaction.js");
var import_checkpoint_contents = require("./checkpoint_contents.js");
var import_signature = require("./signature.js");
var import_checkpoint_summary = require("./checkpoint_summary.js");
class Checkpoint$Type extends import_runtime4.MessageType {
  constructor() {
    super("sui.rpc.v2beta2.Checkpoint", [
      {
        no: 1,
        name: "sequence_number",
        kind: "scalar",
        opt: true,
        T: 4,
        L: 0
      },
      {
        no: 2,
        name: "digest",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "summary", kind: "message", T: () => import_checkpoint_summary.CheckpointSummary },
      { no: 4, name: "signature", kind: "message", T: () => import_signature.ValidatorAggregatedSignature },
      { no: 5, name: "contents", kind: "message", T: () => import_checkpoint_contents.CheckpointContents },
      {
        no: 6,
        name: "transactions",
        kind: "message",
        repeat: 2,
        T: () => import_executed_transaction.ExecutedTransaction
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.transactions = [];
    if (value !== void 0) (0, import_runtime3.reflectionMergePartial)(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional uint64 sequence_number */
        1:
          message.sequenceNumber = reader.uint64().toBigInt();
          break;
        case /* optional string digest */
        2:
          message.digest = reader.string();
          break;
        case /* optional sui.rpc.v2beta2.CheckpointSummary summary */
        3:
          message.summary = import_checkpoint_summary.CheckpointSummary.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.summary
          );
          break;
        case /* optional sui.rpc.v2beta2.ValidatorAggregatedSignature signature */
        4:
          message.signature = import_signature.ValidatorAggregatedSignature.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.signature
          );
          break;
        case /* optional sui.rpc.v2beta2.CheckpointContents contents */
        5:
          message.contents = import_checkpoint_contents.CheckpointContents.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.contents
          );
          break;
        case /* repeated sui.rpc.v2beta2.ExecutedTransaction transactions */
        6:
          message.transactions.push(
            import_executed_transaction.ExecutedTransaction.internalBinaryRead(reader, reader.uint32(), options)
          );
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(
              `Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`
            );
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? import_runtime2.UnknownFieldHandler.onRead : u)(
              this.typeName,
              message,
              fieldNo,
              wireType,
              d
            );
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequenceNumber !== void 0)
      writer.tag(1, import_runtime.WireType.Varint).uint64(message.sequenceNumber);
    if (message.digest !== void 0)
      writer.tag(2, import_runtime.WireType.LengthDelimited).string(message.digest);
    if (message.summary)
      import_checkpoint_summary.CheckpointSummary.internalBinaryWrite(
        message.summary,
        writer.tag(3, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    if (message.signature)
      import_signature.ValidatorAggregatedSignature.internalBinaryWrite(
        message.signature,
        writer.tag(4, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    if (message.contents)
      import_checkpoint_contents.CheckpointContents.internalBinaryWrite(
        message.contents,
        writer.tag(5, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    for (let i = 0; i < message.transactions.length; i++)
      import_executed_transaction.ExecutedTransaction.internalBinaryWrite(
        message.transactions[i],
        writer.tag(6, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    let u = options.writeUnknownFields;
    if (u !== false) (u == true ? import_runtime2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Checkpoint = new Checkpoint$Type();
