// @mysten/sui-grpc/dist/cjs/proto/sui/rpc/v2beta2/executed_transaction.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var executed_transaction_exports = {};
__export(executed_transaction_exports, {
  ExecutedTransaction: () => ExecutedTransaction
});
module.exports = __toCommonJS(executed_transaction_exports);
var import_runtime = require("@protobuf-ts/runtime");
var import_runtime2 = require("@protobuf-ts/runtime");
var import_runtime3 = require("@protobuf-ts/runtime");
var import_runtime4 = require("@protobuf-ts/runtime");
var import_object = require("./object.js");
var import_balance_change = require("./balance_change.js");
var import_timestamp = require("../../../google/protobuf/timestamp.js");
var import_event = require("./event.js");
var import_effects = require("./effects.js");
var import_signature = require("./signature.js");
var import_transaction = require("./transaction.js");
class ExecutedTransaction$Type extends import_runtime4.MessageType {
  constructor() {
    super("sui.rpc.v2beta2.ExecutedTransaction", [
      {
        no: 1,
        name: "digest",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "transaction", kind: "message", T: () => import_transaction.Transaction },
      {
        no: 3,
        name: "signatures",
        kind: "message",
        repeat: 2,
        T: () => import_signature.UserSignature
      },
      { no: 4, name: "effects", kind: "message", T: () => import_effects.TransactionEffects },
      { no: 5, name: "events", kind: "message", T: () => import_event.TransactionEvents },
      {
        no: 6,
        name: "checkpoint",
        kind: "scalar",
        opt: true,
        T: 4,
        L: 0
      },
      { no: 7, name: "timestamp", kind: "message", T: () => import_timestamp.Timestamp },
      {
        no: 8,
        name: "balance_changes",
        kind: "message",
        repeat: 2,
        T: () => import_balance_change.BalanceChange
      },
      {
        no: 10,
        name: "input_objects",
        kind: "message",
        repeat: 2,
        T: () => import_object.Object
      },
      {
        no: 11,
        name: "output_objects",
        kind: "message",
        repeat: 2,
        T: () => import_object.Object
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signatures = [];
    message.balanceChanges = [];
    message.inputObjects = [];
    message.outputObjects = [];
    if (value !== void 0) (0, import_runtime3.reflectionMergePartial)(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional string digest */
        1:
          message.digest = reader.string();
          break;
        case /* optional sui.rpc.v2beta2.Transaction transaction */
        2:
          message.transaction = import_transaction.Transaction.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.transaction
          );
          break;
        case /* repeated sui.rpc.v2beta2.UserSignature signatures */
        3:
          message.signatures.push(
            import_signature.UserSignature.internalBinaryRead(reader, reader.uint32(), options)
          );
          break;
        case /* optional sui.rpc.v2beta2.TransactionEffects effects */
        4:
          message.effects = import_effects.TransactionEffects.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.effects
          );
          break;
        case /* optional sui.rpc.v2beta2.TransactionEvents events */
        5:
          message.events = import_event.TransactionEvents.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.events
          );
          break;
        case /* optional uint64 checkpoint */
        6:
          message.checkpoint = reader.uint64().toBigInt();
          break;
        case /* optional google.protobuf.Timestamp timestamp */
        7:
          message.timestamp = import_timestamp.Timestamp.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.timestamp
          );
          break;
        case /* repeated sui.rpc.v2beta2.BalanceChange balance_changes */
        8:
          message.balanceChanges.push(
            import_balance_change.BalanceChange.internalBinaryRead(reader, reader.uint32(), options)
          );
          break;
        case /* repeated sui.rpc.v2beta2.Object input_objects */
        10:
          message.inputObjects.push(import_object.Object.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated sui.rpc.v2beta2.Object output_objects */
        11:
          message.outputObjects.push(import_object.Object.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(
              `Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`
            );
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? import_runtime2.UnknownFieldHandler.onRead : u)(
              this.typeName,
              message,
              fieldNo,
              wireType,
              d
            );
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.digest !== void 0)
      writer.tag(1, import_runtime.WireType.LengthDelimited).string(message.digest);
    if (message.transaction)
      import_transaction.Transaction.internalBinaryWrite(
        message.transaction,
        writer.tag(2, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    for (let i = 0; i < message.signatures.length; i++)
      import_signature.UserSignature.internalBinaryWrite(
        message.signatures[i],
        writer.tag(3, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    if (message.effects)
      import_effects.TransactionEffects.internalBinaryWrite(
        message.effects,
        writer.tag(4, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    if (message.events)
      import_event.TransactionEvents.internalBinaryWrite(
        message.events,
        writer.tag(5, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    if (message.checkpoint !== void 0) writer.tag(6, import_runtime.WireType.Varint).uint64(message.checkpoint);
    if (message.timestamp)
      import_timestamp.Timestamp.internalBinaryWrite(
        message.timestamp,
        writer.tag(7, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    for (let i = 0; i < message.balanceChanges.length; i++)
      import_balance_change.BalanceChange.internalBinaryWrite(
        message.balanceChanges[i],
        writer.tag(8, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    for (let i = 0; i < message.inputObjects.length; i++)
      import_object.Object.internalBinaryWrite(
        message.inputObjects[i],
        writer.tag(10, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    for (let i = 0; i < message.outputObjects.length; i++)
      import_object.Object.internalBinaryWrite(
        message.outputObjects[i],
        writer.tag(11, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    let u = options.writeUnknownFields;
    if (u !== false) (u == true ? import_runtime2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ExecutedTransaction = new ExecutedTransaction$Type();
