// @mysten/sui-grpc/dist/cjs/proto/sui/rpc/v2beta2/transaction_execution_service.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var transaction_execution_service_exports = {};
__export(transaction_execution_service_exports, {
  ExecuteTransactionRequest: () => ExecuteTransactionRequest,
  ExecuteTransactionResponse: () => ExecuteTransactionResponse,
  TransactionExecutionService: () => TransactionExecutionService,
  TransactionFinality: () => TransactionFinality
});
module.exports = __toCommonJS(transaction_execution_service_exports);
var import_runtime_rpc = require("@protobuf-ts/runtime-rpc");
var import_runtime = require("@protobuf-ts/runtime");
var import_runtime2 = require("@protobuf-ts/runtime");
var import_runtime3 = require("@protobuf-ts/runtime");
var import_runtime4 = require("@protobuf-ts/runtime");
var import_empty = require("../../../google/protobuf/empty.js");
var import_signature = require("./signature.js");
var import_executed_transaction = require("./executed_transaction.js");
var import_field_mask = require("../../../google/protobuf/field_mask.js");
var import_signature2 = require("./signature.js");
var import_transaction = require("./transaction.js");
class ExecuteTransactionRequest$Type extends import_runtime4.MessageType {
  constructor() {
    super("sui.rpc.v2beta2.ExecuteTransactionRequest", [
      { no: 1, name: "transaction", kind: "message", T: () => import_transaction.Transaction },
      {
        no: 2,
        name: "signatures",
        kind: "message",
        repeat: 2,
        T: () => import_signature2.UserSignature
      },
      { no: 3, name: "read_mask", kind: "message", T: () => import_field_mask.FieldMask }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signatures = [];
    if (value !== void 0)
      (0, import_runtime3.reflectionMergePartial)(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional sui.rpc.v2beta2.Transaction transaction */
        1:
          message.transaction = import_transaction.Transaction.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.transaction
          );
          break;
        case /* repeated sui.rpc.v2beta2.UserSignature signatures */
        2:
          message.signatures.push(
            import_signature2.UserSignature.internalBinaryRead(reader, reader.uint32(), options)
          );
          break;
        case /* optional google.protobuf.FieldMask read_mask */
        3:
          message.readMask = import_field_mask.FieldMask.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.readMask
          );
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(
              `Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`
            );
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? import_runtime2.UnknownFieldHandler.onRead : u)(
              this.typeName,
              message,
              fieldNo,
              wireType,
              d
            );
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.transaction)
      import_transaction.Transaction.internalBinaryWrite(
        message.transaction,
        writer.tag(1, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    for (let i = 0; i < message.signatures.length; i++)
      import_signature2.UserSignature.internalBinaryWrite(
        message.signatures[i],
        writer.tag(2, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    if (message.readMask)
      import_field_mask.FieldMask.internalBinaryWrite(
        message.readMask,
        writer.tag(3, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    let u = options.writeUnknownFields;
    if (u !== false) (u == true ? import_runtime2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ExecuteTransactionRequest = new ExecuteTransactionRequest$Type();
class ExecuteTransactionResponse$Type extends import_runtime4.MessageType {
  constructor() {
    super("sui.rpc.v2beta2.ExecuteTransactionResponse", [
      { no: 1, name: "finality", kind: "message", T: () => TransactionFinality },
      { no: 2, name: "transaction", kind: "message", T: () => import_executed_transaction.ExecutedTransaction }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      (0, import_runtime3.reflectionMergePartial)(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional sui.rpc.v2beta2.TransactionFinality finality */
        1:
          message.finality = TransactionFinality.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.finality
          );
          break;
        case /* optional sui.rpc.v2beta2.ExecutedTransaction transaction */
        2:
          message.transaction = import_executed_transaction.ExecutedTransaction.internalBinaryRead(
            reader,
            reader.uint32(),
            options,
            message.transaction
          );
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(
              `Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`
            );
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? import_runtime2.UnknownFieldHandler.onRead : u)(
              this.typeName,
              message,
              fieldNo,
              wireType,
              d
            );
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.finality)
      TransactionFinality.internalBinaryWrite(
        message.finality,
        writer.tag(1, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    if (message.transaction)
      import_executed_transaction.ExecutedTransaction.internalBinaryWrite(
        message.transaction,
        writer.tag(2, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    let u = options.writeUnknownFields;
    if (u !== false) (u == true ? import_runtime2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ExecuteTransactionResponse = new ExecuteTransactionResponse$Type();
class TransactionFinality$Type extends import_runtime4.MessageType {
  constructor() {
    super("sui.rpc.v2beta2.TransactionFinality", [
      {
        no: 1,
        name: "certified",
        kind: "message",
        oneof: "finality",
        T: () => import_signature.ValidatorAggregatedSignature
      },
      {
        no: 2,
        name: "checkpointed",
        kind: "scalar",
        oneof: "finality",
        T: 4,
        L: 0
      },
      { no: 3, name: "quorum_executed", kind: "message", oneof: "finality", T: () => import_empty.Empty }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.finality = { oneofKind: void 0 };
    if (value !== void 0) (0, import_runtime3.reflectionMergePartial)(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* sui.rpc.v2beta2.ValidatorAggregatedSignature certified */
        1:
          message.finality = {
            oneofKind: "certified",
            certified: import_signature.ValidatorAggregatedSignature.internalBinaryRead(
              reader,
              reader.uint32(),
              options,
              message.finality.certified
            )
          };
          break;
        case /* uint64 checkpointed */
        2:
          message.finality = {
            oneofKind: "checkpointed",
            checkpointed: reader.uint64().toBigInt()
          };
          break;
        case /* google.protobuf.Empty quorum_executed */
        3:
          message.finality = {
            oneofKind: "quorumExecuted",
            quorumExecuted: import_empty.Empty.internalBinaryRead(
              reader,
              reader.uint32(),
              options,
              message.finality.quorumExecuted
            )
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(
              `Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`
            );
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? import_runtime2.UnknownFieldHandler.onRead : u)(
              this.typeName,
              message,
              fieldNo,
              wireType,
              d
            );
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.finality.oneofKind === "certified")
      import_signature.ValidatorAggregatedSignature.internalBinaryWrite(
        message.finality.certified,
        writer.tag(1, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    if (message.finality.oneofKind === "checkpointed")
      writer.tag(2, import_runtime.WireType.Varint).uint64(message.finality.checkpointed);
    if (message.finality.oneofKind === "quorumExecuted")
      import_empty.Empty.internalBinaryWrite(
        message.finality.quorumExecuted,
        writer.tag(3, import_runtime.WireType.LengthDelimited).fork(),
        options
      ).join();
    let u = options.writeUnknownFields;
    if (u !== false) (u == true ? import_runtime2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TransactionFinality = new TransactionFinality$Type();
const TransactionExecutionService = new import_runtime_rpc.ServiceType(
  "sui.rpc.v2beta2.TransactionExecutionService",
  [
    {
      name: "ExecuteTransaction",
      options: {},
      I: ExecuteTransactionRequest,
      O: ExecuteTransactionResponse
    }
  ]
);
